pkg/cli/image/strategy/onerror.go

```diff
@@ -162,3 +154,150 @@ func readICSPsFromFile(icspFile string) ([]operatorv1alpha1.ImageContentSourcePo
 	}
 	return []operatorv1alpha1.ImageContentSourcePolicy{*icsp}, nil
 }
+
+type onErrorIDMSStrategy struct {
+	lock sync.Mutex
+
+	alternates            map[reference.DockerImageReference][]reference.DockerImageReference
+	idmsFile              string
+	readIDMSsFromFileFunc readIDMSsFromFileFunc
+}
+
+var _ registryclient.AlternateBlobSourceStrategy = &onErrorIDMSStrategy{}
+
+// NewIDMSOnErrorStrategy returns IDMS alternate strategy which reads alternate
+// sources only after getting an error from the original requested.
+func NewIDMSOnErrorStrategy(file string) registryclient.AlternateBlobSourceStrategy {
+	return &onErrorIDMSStrategy{
+		idmsFile:              file,
+		alternates:            make(map[reference.DockerImageReference][]reference.DockerImageReference),
+		readIDMSsFromFileFunc: readIDMSsFromFile,
+	}
+}
+
+func (s *onErrorIDMSStrategy) FirstRequest(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	return nil, nil
+}
+
+func (s *onErrorIDMSStrategy) OnFailure(ctx context.Context, locator reference.DockerImageReference) (alternateRepositories []reference.DockerImageReference, err error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+	if alternates, ok := s.alternates[locator]; ok {
+		return alternates, nil
+	}
+	alternates, err := s.resolve(ctx, locator)
+	if err != nil {
+		return nil, err
+	}
+	if len(alternates) == 0 {
+		return nil, fmt.Errorf("no alternative image references found for image: %s", locator.String())
+	}
+	s.alternates[locator] = alternates
+	return s.alternates[locator], nil
+}
+
+// resolve gathers possible image sources for a given image
+// gathered from ImageContentSourcePolicy file.
+// Image reference of user-given image may be different from original in case of mirrored images.
+func (s *onErrorIDMSStrategy) resolve(ctx context.Context, imageRef reference.DockerImageReference) ([]reference.DockerImageReference, error) {
+	if len(s.idmsFile) == 0 {
+		return nil, fmt.Errorf("no ImageDigestMirrorSet specified")
+	}
+	klog.V(5).Infof("Reading IDMS from file %s", s.idmsFile)
+	idmsList, err := s.readIDMSsFromFileFunc(s.idmsFile)
+	if err != nil {
+		return nil, err
+	}
+	// always add the original as the first reference
+	imageRefList, err := alternativeImageSourcesIDMS(imageRef, idmsList, false)
+	if err != nil {
+		return nil, err
+	}
+	return imageRefList, nil
+}
+
+// alternativeImageSourcesIDMS returns unique list of DockerImageReference objects from list of ImageDigestMirrorSet objects
+// addSourceAsLastAlternate decides whether the original imageRef is first or the last element in the result
+func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
```

Comment:
Apart from this function, the other changes are mostly copying of ICSP and renaming it to IDMS. But this function, we need to assure that it works properly. @QiWang19 could you please take a look at especially this function?.

Reply:
Sure, I paste the function below.
```go
func alternativeImageSourcesIDMS(imageRef reference.DockerImageReference, idmsList []apicfgv1.ImageDigestMirrorSet, addSourceAsLastAlternate bool) ([]reference.DockerImageReference, error) {
	var imageSources []reference.DockerImageReference
	var addSource bool
	repo := imageRef.AsRepository().Exact()
	for _, idms := range idmsList {
		repoDigestMirrors := idms.Spec.ImageDigestMirrors
		for _, rdm := range repoDigestMirrors {
			...
                        ...
			if imageRef.AsRepository().AsV2() != rdmSourceRef.AsRepository().AsV2() {
				if !isSubrepo(repo, rdm.Source) {
					continue
				}
				suffix = repo[len(rdm.Source):]
			}
			klog.V(5).Infof("%v ImageDigestMirrors source matches given image", imageRef.AsRepository().AsV2())
			// check valid mirrorSourcePolicy
			addSource, err = isAddSource(idmsList, rdm.Source)
			if err != nil {
				return nil, err
			}
			for _, m := range rdm.Mirrors {
				mRef, err := reference.Parse(string(m) + suffix)
				if err != nil {
					return nil, fmt.Errorf("invalid mirror %q: %w", m, err)
				}
				imageSources = append(imageSources, mRef)
				klog.V(5).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
			}
		}
	}
	if addSource && !addSourceAsLastAlternate {
		// prepend imageRef.AsRepository().AsV2()  to the  imageSources
	}
	if addSource && addSourceAsLastAlternate {
		// append imageRef.AsRepository().AsV2() to the imageSources
	}
	uniqueMirrors := make([]reference.DockerImageReference, 0, len(imageSources))
	uniqueMap := make(map[reference.DockerImageReference]bool)
	//  omit the rest of the function ...
	return uniqueMirrors, nil
}
```